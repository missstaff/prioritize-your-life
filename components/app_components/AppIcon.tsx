import React from "react";
import { View } from "react-native";
import { ScaledSheet, s } from "react-native-size-matters";
import Ionicons from "@expo/vector-icons/Ionicons";
import { type IconProps } from "@expo/vector-icons/build/createIconSet";
import { type ComponentProps } from "react";

/**
 * Renders an app icon component.
 *
 * @param style - The style object for the component.
 * @param rest - Additional props for the component.
 * @returns The rendered app icon component.
 */
export function AppIcon({
  style,
  ...rest
}: IconProps<ComponentProps<typeof Ionicons>["name"] | any>): JSX.Element {
  return (
    <View style={[styles.container, style]}>
      <Ionicons size={s(28)} {...rest} />
    </View>
  );
}

const styles = ScaledSheet.create({
  container: {
    alignItems: "center",
    backgroundColor: "transparent",
    borderRadius: 100,
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    padding: s(2.5),
  },
});
