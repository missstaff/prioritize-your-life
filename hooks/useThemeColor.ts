import { useColorScheme } from "react-native";
import { COLORTHEME } from "@/constants/Colors";


export function useThemeColor(
  props: { light?: string; dark?: string },
  colorName: keyof typeof COLORTHEME.light & keyof typeof COLORTHEME.dark
) {
  const theme = useColorScheme() ?? "light";
  const colorFromProps = props[theme];

  if (colorFromProps) {
    return colorFromProps;
  } else {
    return COLORTHEME[theme][colorName];
  }
}
