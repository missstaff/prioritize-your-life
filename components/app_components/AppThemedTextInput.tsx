import React, { useState } from "react";
import {
  StyleProp,
  TextInput,
  TextStyle,
  TouchableOpacity,
  ViewStyle,
  Platform,
} from "react-native";
import { DateTimePickerAndroid } from "@react-native-community/datetimepicker";
import { ScaledSheet, s } from "react-native-size-matters";
import AppIcon from "./AppIcon";
import AppThemedView from "./AppThemedView";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORTHEME } from "@/constants/Colors";
import { TransactionState } from "@/store/transaction/transaction-reducer";

interface AppThemedTextInputProps {
  containerStyle?: StyleProp<ViewStyle>;
  data?: any[] | undefined;
  iconName?: string;
  inputStyle?: StyleProp<TextStyle>;
  keyboardType?: "default" | "numeric";
  placeholder: string;
  secureEntry: boolean;
  value: string;
  checkValue: (value: string) => void;
  setValue: (value: string) => void;
}

const AppThemedTextInput = ({
  containerStyle,
  data,
  inputStyle,
  iconName,
  keyboardType,
  placeholder,
  secureEntry,
  value,
  checkValue,
  setValue,
  ...rest
}: AppThemedTextInputProps) => {
  const textColor = useThemeColor({}, "text");
  let backgroundColor = useThemeColor(
    { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
    "background"
  );

  const [isPasswordVisible, setIsPasswordVisible] = useState(!secureEntry);

  const handleCalendarPress = () => {
    if (Platform.OS === "android") {
      DateTimePickerAndroid.open({
        mode: "date",
        value: new Date(),
        onChange: (event, selectedDate) => {
          if (selectedDate) {
            setValue(formatDate(selectedDate));
          }
        },
      });
    }
  };

  return (
    <AppThemedView style={[styles.inputContainer, containerStyle]}>
      <TextInput
        keyboardType={keyboardType}
        onBlur={(e) => {
          if (value.length > 0) checkValue(value);
        }}
        onChangeText={(text) => {
          setValue(text);
        }}
        placeholder={placeholder}
        placeholderTextColor="#999"
        secureTextEntry={!isPasswordVisible}
        style={[
          styles.input,
          { backgroundColor: backgroundColor, color: textColor },
          inputStyle,
        ]}
        defaultValue={value}
        value={value}
        {...rest}
      />
      {secureEntry && (
        <TouchableOpacity
          onPress={() => setIsPasswordVisible(!isPasswordVisible)}
          testID="passwordVisibilityToggle"
        >
          <AppIcon
            name={!isPasswordVisible ? "eye-off" : "eye"}
            size={s(24)}
            color="#ccc"
          />
        </TouchableOpacity>
      )}

      {iconName && (
        <TouchableOpacity
          disabled={iconName !== "calendar"}
          onPress={() => iconName === "calendar" && handleCalendarPress()}
        >
          <AppIcon name={iconName} size={s(24)} color="#ccc" />
        </TouchableOpacity>
      )}
    </AppThemedView>
  );
};

const styles = ScaledSheet.create({
  inputContainer: {
    alignItems: "center",
    borderColor: "#ccc",
    borderRadius: s(5),
    borderWidth: s(1),
    flexDirection: "row",
    marginBottom: s(10),
    paddingHorizontal: s(10),
    width: "80%",
  },
  input: {
    flex: 1,
    padding: s(10),
  },
});

const formatDate = (date: Date): string => {
  const day = String(date.getDate()).padStart(2, "0");
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const year = date.getFullYear();
  return `${month}/${day}/${year}`;
};

export default AppThemedTextInput;
