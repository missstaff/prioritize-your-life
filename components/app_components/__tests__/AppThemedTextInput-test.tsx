import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react-native';
import { AppThemedTextInput } from '../AppThemedTextInput'; 
import { useThemeColor } from '@/hooks/useThemeColor';
import { AppIcon } from '../AppIcon';

jest.mock('@/hooks/useThemeColor', () => ({
  useThemeColor: jest.fn(),
}));
jest.mock('../AppIcon', () => ({
  AppIcon: jest.fn(() => null),
}));

describe('AppThemedTextInput', () => {
  const mockTextColor = '#000';
  const mockBackgroundColor = '#fff';
  afterEach(() => {
    jest.clearAllMocks();
  });
  
  beforeEach(() => {
    (useThemeColor as jest.Mock).mockImplementation((colors, key) => {
      return key === 'text' ? mockTextColor : mockBackgroundColor;
    });
  });

  it('renders correctly and matches snapshot', () => {
    const { toJSON } = render(
      <AppThemedTextInput
        placeholder="Enter text"
        secureEntry={false}
        value=""
        setValue={() => {}}
        checkValue={() => {}}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('displays the placeholder text', () => {
    render(
      <AppThemedTextInput
        placeholder="Enter text"
        secureEntry={false}
        value=""
        setValue={() => {}}
        checkValue={() => {}}
      />
    );
    expect(screen.getByPlaceholderText('Enter text')).toBeTruthy();
  });

  it('handles text input changes', () => {
    const mockSetValue = jest.fn();
    render(
      <AppThemedTextInput
        placeholder="Enter text"
        secureEntry={false}
        value=""
        setValue={mockSetValue}
        checkValue={() => {}}
      />
    );
    fireEvent.changeText(screen.getByPlaceholderText('Enter text'), 'New text');
    expect(mockSetValue).toHaveBeenCalledWith('New text');
  });

  it('calls checkValue on blur if value is not empty', () => {
    const mockCheckValue = jest.fn();
    render(
      <AppThemedTextInput
        placeholder="Enter text"
        secureEntry={false}
        value="Some value"
        setValue={() => {}}
        checkValue={mockCheckValue}
      />
    );
    fireEvent(screen.getByPlaceholderText('Enter text'), 'blur');
    expect(mockCheckValue).toHaveBeenCalledWith('Some value');
  });

  it('toggles password visibility when eye icon is pressed', () => {
    const { getByTestId } = render(
      <AppThemedTextInput
        placeholder="Password"
        secureEntry={true}
        value=""
        setValue={() => {}}
        checkValue={() => {}}
      />
    );
    // Initially, the icon should be "eye-off"
    expect(getByTestId('passwordVisibilityToggle')).toBeTruthy();
    fireEvent.press(getByTestId('passwordVisibilityToggle'));
    // After pressing, the icon should toggle to "eye"
    expect(AppIcon).toHaveBeenCalledWith(
      expect.objectContaining({ name: 'eye' }),
      expect.anything()
    );
  });

  it('renders an icon if iconName prop is provided', () => {
    render(
      <AppThemedTextInput
        iconName="search"
        placeholder="Search"
        secureEntry={false}
        value=""
        setValue={() => {}}
        checkValue={() => {}}
      />
    );
    expect(AppIcon).toHaveBeenCalledWith(
      expect.objectContaining({ name: 'search' }),
      expect.anything()
    );
  });
});
