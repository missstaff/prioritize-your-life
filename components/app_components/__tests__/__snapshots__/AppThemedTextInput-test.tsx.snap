// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppThemedTextInput renders correctly and matches snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      [
        {
          "alignItems": "center",
          "borderColor": "#ccc",
          "borderRadius": 10.714285714285714,
          "borderWidth": 2.142857142857143,
          "flexDirection": "row",
          "marginBottom": 21.428571428571427,
          "paddingHorizontal": 21.428571428571427,
          "width": "80%",
        },
        undefined,
      ],
    ]
  }
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Enter text"
    placeholderTextColor="#999"
    secureTextEntry={false}
    style={
      [
        {
          "flex": 1,
          "padding": 21.428571428571427,
        },
        {
          "backgroundColor": "#fff",
          "color": "#000",
        },
        undefined,
      ]
    }
    value=""
  />
</View>
`;
